src/controllers/trackController.ts

```diff
@@ -181,15 +188,69 @@ const uploadPatchToS3 = asyncWrapper(
   }
 );
 
+const checkTrackOwner = asyncWrapper(
+  async (req: IRequestWithUser, res: Response, next: NextFunction) => {
+    const track = await Track.findById(req.params.id);
+
+    if (!track) {
+      return next(new AppError('No Track found with that ID ', 404));
+    }
+
+    const isOwner = checkDocumentOwner(track, req.user);
+
+    if (!isOwner)
+      return next(
+        new AppError('You are not authorized to update this document', 401)
+      );
+
+    req.track = track;
+
+    next();
+  }
+);
+
 const createTrack = createOne<ITrack>(Track);
 
 const getAllTracks = getAllEntities<ITrack>(Track);
 
 const getTrack = getOne<ITrack>(Track);
 
-const updateTrack = updateOne<ITrack>(Track, true);
+const updateTrack = asyncWrapper(
+  async (req: IRequestWithUser, res: Response, next: NextFunction) => {
+    let deleteUrls: string[] = [];
+    const track = req.track;
+
+    if (req.body.coverImage && track) deleteUrls.push(track.coverImage);
+    if (req.body.url && track) deleteUrls.push(track.url);
 
-const deleteTrack = deleteOne<ITrack>(Track, true);
+    if (deleteUrls.length > 0) await removesFilesFromS3(deleteUrls);
+
+    const updatedTrack = await Track.findByIdAndUpdate(
+      req.params.id,
+      req.body,
+      { runValidators: true, new: true }
+    );
+
+    res.status(201).json({
+      status: 'success',
+      data: updatedTrack,
+    });
+  }
+);
+
+const deleteTrack = asyncWrapper(
+  async (req: IRequestWithUser, res: Response, next: NextFunction) => {
+    const track = req.track;
+    if (track) await removesFilesFromS3([track.coverImage, track.url]);
+
+    await Track.findByIdAndRemove(req.params.id);
+
+    res.status(204).json({
+      status: 'success',
+      data: null,
+    });
+  }
+);
```

Comment:
Overall really nice work! The only thing I could think of is that maybe remove/update the entity first and then try to delete the previous files, this is in the case for example while updating the track the validators fail and the transaction isn't completed or an error happens the previous track will still exist but without any pictures, I think. I tried testing this by updating a track passing a name with a length of less than 4 at first the image loaded but now it shows this. (I'm not 100% sure if it displays this because of the file that's removed or if it is something else).
![Screenshot 2023-07-25 at 1 23 48 PM](https://github.com/Xaroz/music-on/assets/27152632/634ca784-47d0-49df-acb5-7467816af4e2)

Of course will be nice to heard the other guys thoughts in these.

Reply:
I updated the logic a bit to remove files after the update/delete calls to the DB, also, I put it after sending the response to the user, since the part where it deletes from S3 its not really concern of the user. I tested this and it didnt seem like there was any trouble with any following request being blocked because there's code executing after res.json()

Reply:
@xeno097 @bujosa can correct me and see if its bad practice to add more code after res.json()